nerfstudio/data/datamanagers/full_images_datamanager.py

```diff
@@ -547,5 +563,193 @@ def _undistort_image(
+    # print(cameras[48].reshape(()).width.item()) # outputs 540
+    data["image"] = torch.from_numpy(image)
+    if mask is not None:
+        data["mask"] = mask
+    # dataset.cameras.fx[idx] = float(K[0, 0])
+    # dataset.cameras.fy[idx] = float(K[1, 1])
+    # dataset.cameras.cx[idx] = float(K[0, 2])
+    # dataset.cameras.cy[idx] = float(K[1, 2])
+    # dataset.cameras.width[idx] = image.shape[1]
+    # dataset.cameras.height[idx] = image.shape[0]
+    # dataset.cameras.distortion_params = None
+    if idx == 48:
+        print("ending", camera.width, camera.height)
+    return data
+
+def undistort_view(idx: int, dataset: TDataset, image_type: Literal["uint8", "float32"] = "float32") -> Dict[str, torch.Tensor]:
+    """Undistorts an image to one taken by a linear (pinhole) camera model and returns a new Camera with these updated intrinsics
+    Note: this method does not modify the dataset's attributes at all.
+
+    Returns: The undistorted data (image, depth, mask, etc.) and the new linear Camera object
+    """
+    data = dataset.get_data(idx, image_type)
+    camera = dataset.cameras[idx].reshape(())
+    assert data["image"].shape[1] == camera.width.item() and data["image"].shape[0] == camera.height.item(), (
+        f'The size of image ({data["image"].shape[1]}, {data["image"].shape[0]}) loaded '
+        f'does not match the camera parameters ({camera.width.item(), camera.height.item()}), idx = {idx}'
+    )
+    if camera.distortion_params is None or torch.all(camera.distortion_params == 0):
+        return data
+    K = camera.get_intrinsics_matrices().numpy()
+    distortion_params = camera.distortion_params.numpy()
+    image = data["image"].numpy()
+    K, image, mask = _undistort_image(camera, distortion_params, data, image, K)
+    data["image"] = torch.from_numpy(image)
+    if mask is not None:
+        data["mask"] = mask
+    
+    # create a new Camera
+    new_camera = Cameras(
+        camera_to_worlds=camera.camera_to_worlds.unsqueeze(0),
+        fx=torch.Tensor([[float(K[0, 0])]]),
+        fy=torch.Tensor([[float(K[1, 1])]]),
+        cx=torch.Tensor([[float(K[0, 2])]]),
+        cy=torch.Tensor([[float(K[1, 2])]]),
+        width=torch.Tensor([[image.shape[1]]]).to(torch.int32),
+        height=torch.Tensor([[image.shape[0]]]).to(torch.int32),
+    )
+    return data, new_camera
+
+import math
+class ImageBatchStream(torch.utils.data.IterableDataset):
+    """
+    A wrapper of InputDataset that outputs undistorted full images and cameras. This makes the
+    datamanager more lightweight since we don't have to do generate rays. Useful for full-image
+    training e.g. rasterization pipelines
+    """
+
+    def __init__(
+        self,
+        datamanager_config: DataManagerConfig,
+        input_dataset: TDataset,
+        device,
+    ):
+        self.config = datamanager_config
+        self.input_dataset = input_dataset
+        self.device = device
+
+    def __iter__(self):
```

Comment:
iteresting so this works fast enough to undistort and not need caching of the undistorted image?  curious what CPU / GPU combo and number of workers was good here.  

or maybe it does slow training a bit, but rather *it works without OOM* on bigger datasets :)

Reply:
Yeah it does slow training quite a bit. With 1 worker, training time went from 9 minutes -> 18 minutes (with downsampling to a resolution inside 1600x1600)

It also depends on what type of undistortion is occuring. I only have tested this with datasets of CAMERA.PERSPECTIVE types, but with fisheye and equirectangular camera models, it will definitely take longer to undistort this, will do some further benchmarks on this!

Reply:
ah interesting!  2x is not that bad, given that this change makes training _feasible_ when low RAM.

maybe caching the undistorted image could help then?  so in your other code, you try to cache the jpeg bytes which helps, and clearly here you can't necessarily cache the compressed image and so you have to cache the undistorted image.  

all that said, it's very nice to even have training _enabled_ even if u don't implement / test caching